### ISSUE TITLE
Type tracking conditionally extended empty tuple

### ISSUE THREAD

**Issue by Fatsie** [microsoft/pylance-release#1359]
## Environment data

-   Language Server version: 2021.5.4
-   OS and version: Ubuntu 20.04
-   Python version (& distribution if applicable, e.g. Anaconda): conda 3.7.10

## Reproducer & Actual behaviour

I reduced my problem to following code that can reproduce it:
```python
from typing import Tuple

def f(n: int, check: bool):
    t: Tuple[Tuple[int, float], ...] = tuple()
    for _ in range(n):
        t += ((0, 0.0),)
        if check:
            t += ((1, 1.0), (2, 2.0))

    for (i, f) in t:
        print(i)
```

This give the following error on the second for line:
```
{
	"resource": "/home/verhaegs/eda/test/test_emptytuple.py",
	"owner": "_generated_diagnostic_collection_name_#2",
	"code": {
		"value": "reportGeneralTypeIssues",
		"target": {
			"$mid": 1,
			"external": "https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules",
			"path": "/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md",
			"scheme": "https",
			"authority": "github.com",
			"fragment": "diagnostic-severity-rules"
		}
	},
	"severity": 8,
	"message": "\"object\" is not iterable\n  \"__iter__\" method not defined",
	"source": "Pylance",
	"startLineNumber": 10,
	"startColumn": 9,
	"endLineNumber": 10,
	"endColumn": 15
}
```

Problem seems related to the fact that on the empty tuple() assignment the type of t is `tuple[_T_co@tuple]` and not the explicitly specified type. When I move the first t += statement outside the loop all is honky dory.

## Expected behaviour

No typing error given.

**Comment by erictraut** [comment]
Thanks for the bug report. The problem appears to be with the initial assignment of `t` — an interaction between the `tuple()` constructor call and the type annotation. I'll investigate further. In the meantime, you can work around the problem by replacing the call to the constructor (`tuple()`) with an empty tuple expression (`()`).

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.6.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202160-2-june-2021
